fetcher/fetcher.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fetcher
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/keys"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// Fetcher retrieves values on-the-fly and ensures that
+// a value is only fetched from disk once. Subsequent
+// requests can be retrieved from cache.
+type Fetcher struct {
+	im        state.Immutable
+	cacheLock sync.RWMutex
+	cache     map[string]*fetchData
+
+	keysToFetch map[string]*keyData        // Number of txns waiting for a key
+	txnsToFetch map[ids.ID]*sync.WaitGroup // Number of keys a txn is waiting on
+	keyLock     sync.RWMutex
+	txnLock     sync.Mutex
+
+	stopOnce  sync.Once
+	stop      chan struct{}
+	err       error
+	fetchable chan *task
+	l         sync.Mutex
+	wg        sync.WaitGroup
+
+	completed int
+	numTxs    int
+}
+
+// Data to insert into the cache
+type fetchData struct {
+	v      []byte
+	exists bool
+	chunks uint16
+}
+
+// Data pertaining to whether a key was fetched or which txns are waiting
+type keyData struct {
+	fetched bool
+	queue   []ids.ID
+}
+
+// Holds the information that a worker needs to fetch values
+type task struct {
+	ctx context.Context
+	id  ids.ID
+	key string
+}
+
+// New creates a new [Fetcher]
+func New(numTxs int, concurrency int, im state.Immutable) *Fetcher {
+	f := &Fetcher{
+		im:          im,
+		cache:       make(map[string]*fetchData, numTxs),
+		keysToFetch: make(map[string]*keyData),
+		txnsToFetch: make(map[ids.ID]*sync.WaitGroup, numTxs),
+		fetchable:   make(chan *task),
+		stop:        make(chan struct{}),
+		numTxs:      numTxs,
+	}
+	for i := 0; i < concurrency; i++ {
+		f.wg.Add(1)
+		go f.runWorker()
+	}
+	return f
+}
+
+// Workers fetch individual keys
+func (f *Fetcher) runWorker() {
+	defer f.wg.Done()
+
+	for {
+		select {
+		case t, ok := <-f.fetchable:
+			if !ok {
+				return
+			}
+
+			// Allows concurrent reads to cache.
+			if exists := f.has(t.key); exists {
+				f.updateDependencies(t.key)
+				continue
+			}
+
+			// Doesn't exist in cache, check if we were the first to
+			// request the key.
+			f.keyLock.RLock()
+			var firstTx ids.ID
+			if len(f.keysToFetch[t.key].queue) > 0 {
+				firstTx = f.keysToFetch[t.key].queue[0]
+			}
+			f.keyLock.RUnlock()
+			if firstTx != t.id {
+				// Fetch other keys instead
+				continue
+			}
+
+			// Fetch from disk once if key isn't already in cache
+			v, err := f.im.GetValue(t.ctx, []byte(t.key))
+			if errors.Is(err, database.ErrNotFound) {
+				f.updateCache(t.key, nil, false, 0)
+				f.updateDependencies(t.key)
+				continue
+			} else if err != nil {
+				f.stopOnce.Do(func() {
+					f.err = err
+					close(f.stop)
+				})
+				return
+			}
+
+			// We verify that the [NumChunks] is already less than the number
```

Comment:
this comment doesn't match what we do below (where we check `ok`)?

Reply:
Oops I borrowed it when copying over the logic from https://github.com/ava-labs/hypersdk/blob/main/chain/processor.go#L94-L102 , however, I removed the comment since I think the `ok` check and what follows is clear